packages/fast-tooling-react/src/form/utilities/navigation.ts

```diff
@@ -0,0 +1,381 @@
+import { cloneDeep, get, memoize } from "lodash-es";
+import { MemoizedFunction } from "lodash";
+import {
+    getDataLocationsOfChildren,
+    getPartialData,
+    mapSchemaLocationFromDataLocation,
+    normalizeDataLocation,
+} from "../../data-utilities/location";
+import { FormChildOptionItem } from "../form";
+import { getReactDefaultChildren, propsKeyword } from "./form";
+import { reactChildrenStringSchema } from "../form/form-item.children.text";
+
+export interface NavigationItemConfig {
+    dataLocation: string;
+    schemaLocation: string;
+    schema: unknown;
+    data: unknown;
+    default: unknown;
+}
+
+export interface NavigationItem {
+    /**
+     * A string using lodash path syntax for the path where this data is located
+     */
+    dataLocation: string;
+
+    /**
+     * A string using lodash path syntax for the path where this schema is located
+     */
+    schemaLocation: string;
+
+    /**
+     * The data
+     */
+    data: unknown;
+
+    /**
+     * The schema or schema partial
+     */
+    schema: unknown;
+
+    /**
+     * The default data if default data exists
+     */
+    default: unknown;
+
+    /**
+     * The title
+     */
+    title: string;
+}
+
+export interface NavigationMap {
+    [key: string]: NavigationItem;
+}
+
+export interface Component {
+    dataLocation: string;
+    schema: unknown;
+}
+
+export interface NavigationConfig {
+    dataLocation: string;
+    data: unknown | void;
+    schema: unknown;
+    childOptions: FormChildOptionItem[];
+}
+
+class Navigation {
+    private navigationMap: NavigationMap;
+
+    private config: NavigationConfig;
+
+    private dataLocation: string;
+
+    private reactChildrenDataLocations: string[];
+
+    private resolveNormalizeDataLocation: ((dataLocation: string) => string) &
+        MemoizedFunction;
+
+    private resolveDataLocations: ((dataLocation: string) => string[]) & MemoizedFunction;
```

Comment:
Do you need to type this as a `MemoizedFunction`? A memorzed function has the same signature as the callback - not sure why this needs extra type information.

Reply:
Yep, turns out I didn't need it, not sure why it was erroring elsewhere when I tried to type it without using that.